File: 2018/ruby/day03.rb
input = File.readlines("/Users/hhh/JungleGym/advent_of_code/2018/inputs/day3.in")
# input = File.readlines("/Users/hhh/JungleGym/advent_of_code/2018/inputs/day3test.in")

claims = {}

input.map do |line|
  /(?<id>\d+) @ (?<x>\d+),(?<y>\d+): (?<w>\d+)x(?<l>\d+)/ =~ line
  claim = {
    x: x.to_i,
    y: y.to_i,
    width: w.to_i,
    length: l.to_i,
  }
  claims[id] = claim
end

cloth = Hash.new{ |h, k| h[k] = Hash.new { |m, n| m[n] = [] } }
overlaps = 0

claims.each do |id, claim|
  x, y, w, l = claim[:x], claim[:y], claim[:width], claim[:length]
  (x...x+w).each do |i|
    (y...y+l).each do |j|
      if !cloth[j][i].empty?
        cloth[j][i].each do |claim_id|
          claims[claim_id][:overlapped] = true
          claims[id][:overlapped] = true
        end
      end

      cloth[j][i] << id
    end
  end
end

cloth.each do |col, row|
  row.each do |pt, val|
    if val.size > 1
      overlaps += 1
    end
  end
end

p "Part 1: #{overlaps}"

claim_id, _ = claims.find do |id, claim|
  !claim[:overlapped]
end

p "Part 2: #{claim_id}"

File: 2018/go/day03/day03.go
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
)

func Populate(claims map[int]Claim) (fabric [1000][1000][]int) {
	for id, claim := range claims {
		for i := claim.X; i < claim.X + claim.Width; i++ {
			for j := claim.Y; j < claim.Y + claim.Length; j++ {
				fabric[j][i] = append(fabric[j][i], id)
			}
		}
	}
	return
}

func Overlap(claims map[int]Claim) int {
	overlaps := 0
	fabric := Populate(claims)

	for _, row := range fabric {
		for _, col := range row {
			if len(col) > 1 {
				overlaps++
			}
		}
	}
	return overlaps
}


func main() {
	input, err := ioutil.ReadFile("/Users/hhh/JungleGym/advent_of_code/2018/inputs/day3.in")
	if err != nil {
		panic(err)
	}
	claims := formatInput(string(input))

	fmt.Printf("Day 3 Part 1: %+v\n", Overlap(claims))
}

type Claim struct {
	X int
	Y int
	Width int
	Length int
	Overlap bool
}

func formatInput(input string) map[int]Claim {
	lines := strings.Split(input, "\n")
	claims := map[int]Claim{}
	for _, line := range lines {

		var id, x, y, w, l int
		// #1 @ 749,666: 27x15
		fmt.Sscanf(line, "#%d @ %d,%d: %dx%d", &id, &x, &y, &w, &l)
		c := Claim{x, y, w, l, false}

		claims[id] = c
	}
	return claims
}
